test/helpers/integration.rb

```diff
@@ -139,4 +138,151 @@ def get_worker_pids(phase = 0, size = WORKERS)
     end
     pids.map(&:to_i)
   end
+
+  # used to define correct 'refused' errors
+  def thread_run_refused(unix: false)
+    if unix
+      [Errno::ENOENT, IOError]
+    else
+      DARWIN ? [Errno::ECONNREFUSED, Errno::EPIPE, EOFError] :
+        [Errno::ECONNREFUSED]
+    end
+  end
+
+  def cli_pumactl(argv, unix: false)
+    if unix
+      pumactl = IO.popen("#{BASE} bin/pumactl -C unix://#{@control_path} -T #{TOKEN} #{argv}", "r")
+    else
+      pumactl = IO.popen("#{BASE} bin/pumactl -C tcp://#{HOST}:#{@control_tcp_port} -T #{TOKEN} #{argv}", "r")
+    end
+    @ios_to_close << pumactl
+    Process.wait pumactl.pid
+    pumactl
+  end
+
+  def hot_restart_does_not_drop_connections(num_threads: 1, ttl_requests: 500)
+    skipped = true
+    skip_on :jruby, suffix: <<-MSG
+ - file descriptors are not preserved on exec on JRuby; connection reset errors are expected during restarts
```

Comment:
Seems like the behavior on JRuby should be the same as on Windows: Since `exec` doesn't preserve file descriptors, puma has to unbind-and-rebind the listening socket, so we lose some connections during restart.

Is there any reason why we run the test on Windows (with the expectation that the number of connection reset errors per thread is bound from above by the number of restarts), but skip the test on JRuby? Seems like we can make the same decision in both places.

Either we can skip on both platforms (connection reset errors are expected on both, so maybe not worth having explicit tests for), or we can run it on both.

Reply:
I think I lean towards just skipping it on both. Since we _expect_ some connection reset errors on Windows, when this test runs on Windows, it's basically just testing that we can make requests to a running Puma server on that platform and get a successful response; it doesn't seem like it adds a lot of value on top of existing tests.

Reply:
@cjlarose

I don't have strong feelings either way.  I doubt there are many publicly facing Windows Puma instances.

Running it locally, I found the reset errors were bounded and a very small percentage of total requests, and I think resets will normally be retried by most clients.  Hence, I including Windows in the tests since if the tests start failing, something has changed for the worse.  Again, not bothered by removing Windows tests.

Re JRuby and/or TruffleRuby, I'm not sure if I tried them, especially when the code was close to finished...

Reply:
I just added JRuby in my fork, it definitely did not pass the new tests.  See:
https://github.com/MSP-Greg/puma/actions/runs/302733028

Also, see comment below ('I hate intermittent tests') for more info.

Reply:
These tests are somewhat dependent on timing.  Restarting on JRuby probably takes a bit longer than MRI Rubies, and the gap during restart will be difficult to account for.

Reply:
Makes sense. In my fork, I had changed the test to run for a max number of seconds instead of doing a particular number of requests. https://github.com/cjlarose/puma/commit/691a5e429ecafbfefc820c32142cb27b1d0120b5

Just something to consider

Reply:
I kind of felt that it was working, and along with more runs to see if they're stable, some of the quantities in the tests may be able to change.

I started with 'single' tests, added the threading, then added the 'cluster' tests.  I added the debug output at the end of the test log, and noticed that on (probably) the threaded cluster test, all the requests were being processed before the first restart.  That's when I kicked the counts up, and later added the small delay in `hello_with_delay.ru`.  

They're all finishing fairly quickly, and seeing the cluster threaded test complete 3k requests in less than 3 seconds seemed like it could stay that way even though the 3k number could probably drop to 2k.

After seeing some runs during the busy CI time (Thurs & Fri late afternoon), we might be able to add a 'something slowed Puma down' assert.  Really don't have anything like that in the test suite.

BTW, thanks for taking a look at it, and your PR's...